src/gwt/src/org/rstudio/studio/client/workbench/views/terminal/xterm/XTermNative.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * XTermNative.java
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+package org.rstudio.studio.client.workbench.views.terminal.xterm;
+
+import org.rstudio.core.client.CommandWithArg;
+
+import com.google.gwt.core.client.JavaScriptObject;
+import com.google.gwt.dom.client.Element;
+
+/**
+ * <code>JavaScriptObject</code> wrapper for xterm.js
+ */
+public class XTermNative extends JavaScriptObject
+{
+   // Required by JavaScriptObject subclasses
+   protected XTermNative()
+   {
+   }
+   
+   /**
+    * Remove event handlers and detach from parent node.
+    */
+   // TODO: (gary) should be calling this
+   public final native void destroy() /*-{
+      this.destroy();
+   }-*/;
+   
+   /**
+    * Fit the terminal to available space.
+    */
+   public final native void fit() /*-{
+      this.fit();
+   }-*/;
+
+   /**
+    * Write a line of output to the terminal, appending CR/LF.
+    * @param data String to write
+    */
+   public final native void writeln(String data) /*-{
+      this.writeln(data);
+   }-*/;
+   
+   /**
+    * Write text to the terminal.
+    * @param data String to write
+    */
+   public final native void write(String data) /*-{
+      this.write(data);
+   }-*/;
+  
+   /**
+    * Compute and return available dimensions for terminal.
+    * @return Visible number of columns and rows
+    */
+   public final native XTermDimensions proposeGeometry() /*-{
+      return this.proposeGeometry();
+   }-*/;
+   
+   public final native void focus() /*-{
+      this.focus(); 
+   }-*/; 
+  
+   public final native void blur() /*-{
+      this.blur(); 
+   }-*/; 
+  
+   public final native void scrollDisp(int lineCount) /*-{
+      this.scrollDisp(lineCount, false);
+   }-*/;
+  
+   public final native void scrollPages(int pageCount) /*-{
+      this.scrollPages(pageCount);
+   }-*/;
+   
+   public final native void scrollToTop() /*-{
+      this.scrollToTop();
+   }-*/;
+   
+   public final native void scrollToBottom() /*-{
+      this.scrollToBottom();
+   }-*/;
+
+   public final native void onData(CommandWithArg<String> command) /*-{
+      this.handler = 
+         $entry(function(data) {
+            command.@org.rstudio.core.client.CommandWithArg::execute(Ljava/lang/Object;)(data);
+         });
+   }-*/;
+   
+   public final native void onTerminalData(CommandWithArg<String> command) /*-{
+      this.handler = 
```

Comment:
What's the difference between `onData` and `onTerminalData` (their bodies look the same)? Also, is it possible to stack these? (i.e. can you ever have multiple handlers?)